File: 5.最长回文子串.java
/*
 * @lc app=leetcode.cn id=5 lang=java
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        int length = s.length();
        String ans = "";

        boolean[][] p = new boolean[length][length];
        for(int len = 1 ; len <= length; len ++){
            for(int begin = 0; begin < length; begin ++){
                int end =begin + len - 1;
                if(end >= length){ continue; }
                p[begin][end] = (len == 1 || len ==2 || p[begin + 1][end - 1]) && (s.charAt(begin) == s.charAt(end));
                ans = (p[begin][end] && ans.length() < len)  ? s.substring(begin, end + 1) : ans;
            }
        }
        return ans;
    }
}
// @lc code=end

File: 5.最长回文子串.go
/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */

// @lc code=start
func longestPalindrome(s string) string {
	var res = ""
	var length = len(s)
	var dp = make([][]bool, length)
	for i := 0; i < length; i++ {
		dp[i] = make([]bool, length)
	}
	for l := 1; l <= length; l++ {
		for i := 0; i < length - l + 1; i ++ {
			if l == 1 {
				dp[i][i] = true
			} else if l == 2 {
				dp[i][i + 1] = s[i] == s[i + 1]
			} else if s[i] == s[i + l - 1]{
				dp[i][i + l - 1] = dp[i + 1][i + l - 2]
			}
			if dp[i][i + l - 1] {
				res = s[i:i + l]
			}
		}
	}
	return res
}
// @lc code=end

